### ISSUE TITLE
Add REST API endpoint for search

### ISSUE THREAD

**Issue by mikkonie** [bihealth/sodar-core#611]
Discussed this with @holtgrewe. Updated in 2024-04.

## Requirements

- Should support all search features the UI does
- Should return found objects in standard serializer format where applicable
- **Update:** I've identified several problems in doing this which we should discuss
    * API search doesn't really map with the current UI based search
        + The plugin search API returns items based on category and item type, not object type
        + We may return results which are of mixed object types (filesfolders) or not database objects at all (iRODS files in sodar/samplesheets)
    * We need to somehow serialize results from apps in search
        + If using "centralized" search via one entry point, we'll somehow need to know how to serialize the objects. This would at the very least require a "get serialized object" method implementation in the plugin, OR refactoring the entire `search()` method for API use
    * Returning multiple types of results from multiple apps in one entrypoint response causes problems
        + Can't nicely add things such as pagination without being convoluted
        + Different object types as mentioned above
    * Tl;dr: plugging REST API search to the current UI search is not feasible without a lot of breaking changes to the plugin UI and might not be feasible at all
- **TBD:** We should decide on the desired scheme for REST API search. I can identify two possible approaches:
    * Centralized search API endpoint similar to UI
        + We need to solve problems detailed above
    * OR, model-specific search/filtering API views for each app
        + This would ensure known results from each query, mapping to a specific object type
        + We could provide some base helper view/mixin to facilitate this
        + Of course, if the use case is to retrieve search results for all apps/types like in the UI, this would not be a good solution (needs multiple queries to different API views etc)
- We should determine currently known use cases to know which approach to go for

I'm removing this issue from current milestones until we've had the opportunity to brainstorm this and solve the aforementioned questions.

**Comment by mikkonie** [comment]
I looked into this and identified several open questions which should be solved together with potential users of this API. Hence, I'm removing this issue from current milestones for now, until we've had the opportunity to brainstorm this and come to decisions. See open questions in the initial comment.
